### PR TITLE
hcm classifier

### PR DESCRIPTION
**Pull request by SammyRamone**
## Proposed changes
Enables using more sophisticated classifiers in HCM falling detection.
Needs to be refactored and have some config value to switch between different classifiers.

## Necessary checks
- [x] Update package version
- [x] Run `catkin build`
- [x] Write documentation
- [x] Create issues for future work
- [ ] Test on your machine
- [x] Test on the robot
- [x] Put the PR on our Project board

### CODE REVIEW

**Comment by SammyRamone** [comment]
currently the scripts for recording data and training the classifier are not in the HCM package. What is your oppinion? Should they be here or rather in a separate git?

**Comment by timonegk** [comment]
Do the files include larger training sets? If yes, I would prefer to have them in a separate git because they are not needed on the robot or locally to execute the falling detection. But if it's just a few small scripts I don't see an issue with having them in this repo.

**Comment by SammyRamone** [comment]
> Do the files include larger training sets? If yes, I would prefer to have them in a separate git because they are not needed on the robot or locally to execute the falling detection. But if it's just a few small scripts I don't see an issue with having them in this repo.

its around 100Mb, so not as large as our image datasets, but still a bit much for a repo. Furthermore, those datasets may change, since you may want to rerecord them for different grass or after changing something on the hardware.
The scripts to record data, label it and to train the classifier are not large.
I could add the scripts to this repo and put the current dataset in the cloud + link in readme.

File: bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py [hunk 1]
```diff
@@ -86,20 +91,23 @@ def __init__(self):
         self.stand_up_right_animation = rospy.get_param("hcm/animations/stand_up_right")
 
         # motors
-        self.last_motor_goal_time = rospy.Time.now() # initilize with current time, or motors will be turned off on start
+        # initialize with current time, or motors will be turned off on start
+        self.last_motor_goal_time = rospy.Time.now()
         self.last_motor_update_time = rospy.Time.from_sec(0)
         self.motor_timeout_duration = rospy.get_param("hcm/motor_timeout_duration")
         self.motor_off_time = rospy.get_param("hcm/motor_off_time")
+        # todo rename in current joint state
```
Comment (timonegk) [inline bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py:99]: There is still a todo here

File: bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py [hunk 2]
```diff
@@ -114,9 +122,36 @@ def __init__(self):
...
-        self.last_kick_feedback = None      # type: rospy.Time
+        self.last_kick_feedback = None  # type: rospy.Time
+
+        # direct messages for falling classier
+        # todo needs refactoring
+
+        rospack = rospkg.RosPack()
+        rospack.list()
+        path = rospack.get_path('bitbots_hcm')
+        self.classifier = FallClassifier(path + "/src/bitbots_hcm/classifier/")
```
Comment (timonegk) [inline bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py:133]: The second argument, smooth_threshold, should probably come from a parameter.

**Review by timonegk — CHANGES_REQUESTED** [review]

File: bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py [hunk 3]
```diff
@@ -114,9 +121,37 @@ def __init__(self):
...
-        self.last_kick_feedback = None      # type: rospy.Time
+        self.last_kick_feedback = None  # type: rospy.Time
+
+        # direct messages for falling classier
+        # todo needs refactoring
+
+        rospack = rospkg.RosPack()
+        rospack.list()
+        path = rospack.get_path('bitbots_hcm')
+        smooth_threshold = rospy.get_param("/smooth_threshold", 10)
```
Comment (timonegk) [inline bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py:132]: The parameter name should be something like `hcm/smooth_threshold` and should be added to the config file.

**Comment by SammyRamone** [comment]
I added the missing scripts for training, as well as documentation with a link to the existing datasets

**Comment by SammyRamone** [comment]
@Flova since you did some work on the original fall detection, it would be nice if you have a look at it before I merge it

File: bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py (hunk 3 above)
Comment (timonegk) [reply bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py:132]: Did you forget to push something or why did you resolve this?

File: bitbots_dynamic_kick/src/kick_ik.cpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ void KickIK::reset() {
    * since we are using a gradient based method. Otherwise, the
    * first step will be not correct */
   std::vector<std::string> names_vec = {"LHipPitch", "LKnee", "LAnklePitch", "RHipPitch", "RKnee", "RAnklePitch"};
-  std::vector<double> pos_vec = {0.7, -1.0, -0.4, -0.7, 1.0, 0.4};
+  std::vector<double> pos_vec = {0.7, 1.0, -0.4, -0.7, -1.0, 0.4};
```
Comment (Flova) [inline bitbots_dynamic_kick/src/kick_ik.cpp:24]: I this part of this pull request?

File: bitbots_dynup/src/dynup_ik.cpp [hunk 5]
```diff
@@ -16,7 +16,7 @@ void DynupIK::reset() {
    * since we are using a gradient based method. Otherwise, the
    * first step will be not correct */
   std::vector<std::string> names_vec = {"LHipPitch", "LKnee", "LAnklePitch", "RHipPitch", "RKnee", "RAnklePitch"};
-  std::vector<double> pos_vec = {0.7, -1.0, -0.4, -0.7, 1.0, 0.4};
+  std::vector<double> pos_vec = {0.7, 1.0, -0.4, -0.7, -1.0, 0.4};
```
Comment (Flova) [inline bitbots_dynup/src/dynup_ik.cpp:19]: Also this

File: bitbots_hcm/src/bitbots_hcm/training/train.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    classifier = sklearn.svm.SVC(C=study.best_params['svc_c'], gamma="auto", kernel=study.best_params['kernel'],
+                                 degree=study.best_params['degree'])
+elif args.gpc:
+    classifier = GaussianProcessClassifier(multi_class=study.best_params['multi_class'])
+elif args.bb:
+    classifier = FallChecker(thresh_gyro_front=study.best_params['thresh_gyro_front'],
+                             thresh_gyro_side=study.best_params['thresh_gyro_side'],
+                             thresh_orient_front=study.best_params['thresh_orient_front'],
+                             tresh_orient_side=study.best_params['tresh_orient_side'])
+classifier.fit(x_scaled, y)
```
Comment (Flova) [inline bitbots_hcm/src/bitbots_hcm/training/train.py:218]: In the case of the bb classifier this would fail, right? Because the `FallChecker` class has no methods such as `.fit()` or `.predict()`. Or is there something I overlooked?

**Review by Flova — CHANGES_REQUESTED** [review]

**Review by timonegk — COMMENTED** [review]
change `tresh` to `thresh`

File: bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py (hunk 3 above)
Comment (SammyRamone) [reply bitbots_hcm/src/bitbots_hcm/hcm_dsd/hcm_blackboard.py]: should be fixed now

File: bitbots_dynamic_kick/src/kick_ik.cpp (hunk 4 above)
Comment (SammyRamone) [reply bitbots_dynamic_kick/src/kick_ik.cpp:24]: this is a fix to adapt kick to new robot model (knee joints changed and therefore the sign changed). was not supposed to be in this PR but needs to be on the master never the less. I would just keep it here, to spare additional work for another PR.

File: bitbots_dynup/src/dynup_ik.cpp (hunk 5 above)
Comment (SammyRamone) [reply bitbots_dynup/src/dynup_ik.cpp:19]: same as above

File: bitbots_hcm/src/bitbots_hcm/training/train.py (hunk 6 above)
Comment (SammyRamone) [reply bitbots_hcm/src/bitbots_hcm/training/train.py:218]: Good catch! I actually forgot to commit one file, should be fixed now.

**Comment by SammyRamone** [comment]
I hope I fixed all issues that you mentioned, please have another look

File: bitbots_hcm/src/bitbots_hcm/fall_checker.py [hunk 7]
```diff
@@ -1,112 +1,71 @@
...
-                self.falling_threshold_orientation_front_back == 0 or \
-                self.falling_threshold_orientation_left_right == 0:
-            return
         # setting the fall quantification function
         x_fall_quantification = self.calc_fall_quantification(
-            self.falling_threshold_orientation_left_right,
-            self.falling_threshold_front,
-            euler[0] - self.pitch_offset,
+            self.thresh_orient_side,
+            self.thresh_gyro_front,
```
Comment (Flova) [inline bitbots_hcm/src/bitbots_hcm/fall_checker.py:41]: These seem switched. This is no error of this PR, but I think the `x_quantification` should use `thresh_orient_side` and `thresh_gyro_side`.

File: bitbots_hcm/src/bitbots_hcm/fall_checker.py [hunk 8]
```diff
@@ -1,112 +1,71 @@
...
+            euler[0],
             not_much_smoothed_gyro[0])
 
         y_fall_quantification = self.calc_fall_quantification(
-            self.falling_threshold_orientation_front_back,
-            self.falling_threshold_side,
+            self.thresh_orient_front,
+            self.thresh_gyro_side,
```
Comment (Flova) [inline bitbots_hcm/src/bitbots_hcm/fall_checker.py:47]: Same here

File: bitbots_hcm/src/bitbots_hcm/fall_checker.py [hunk 9]
```diff
@@ -120,28 +79,17 @@ def calc_fall_quantification(self, falling_threshold_orientation, falling_thresh
...
-        if smooth_accel[0] > 7:
-            rospy.loginfo("FALLEN TO THE BACK")
-            return self.BACK
-
-        if smooth_accel[1] > 7:
-            rospy.loginfo("FALLEN TO THE RIGHT")
-            return self.RIGHT
+    def fit(self, x, y):
+        # we have to do nothing, as we are not actually fitting any model
+        pass
```
Comment (Flova) [inline bitbots_hcm/src/bitbots_hcm/fall_checker.py:84]: Maybe we should print a log message in case someone tries to train this classifier.

**Comment by SammyRamone** [comment]
incooperated the wanted changes, I will test it in the next days

**Review by Flova — APPROVED** [review]

File: bitbots_hcm/src/bitbots_hcm/fall_checker.py [hunk 10]
```diff
@@ -94,3 +94,29 @@ def predict(self, x):
...
         return y
+
+    def check_fallen(self, smooth_accel, not_much_smoothed_gyro):
+        """Check if the robot has fallen and is lying on the floor. Returns animation to play, if necessary."""
+        # Checks if the robot is still moving.
+        if any(abs(n) >= 0.1 for n in not_much_smoothed_gyro):
+            return None
+
+        # Decides which side is facing downwards.
+        if smooth_accel[0] < -7:
```
Comment (NFiedler) [inline bitbots_hcm/src/bitbots_hcm/fall_checker.py]: Why is this value not configurable?

Comment (SammyRamone) [reply bitbots_hcm/src/bitbots_hcm/fall_checker.py]: because it always worked perfect and we never needed to change it

**Comment by SammyRamone** [comment]
also fixes 
https://github.com/bit-bots/bitbots_motion/issues/151

**Comment by SammyRamone** [comment]
tested it on the robot and works now. @timonegk and @NFiedler  please review

**Review by timonegk — APPROVED** [review]

**Review by NFiedler — APPROVED** [review]

File: bitbots_hcm/config/hcm_wolfgang.yaml [hunk 11]
```diff
@@ -37,14 +37,14 @@ hcm:
...
-  falling_threshold_orientation_left_right: 60 # > Point of no return in degrees
-  falling_threshold_orientation_front_back: 45 # < Point of no return in degrees
-  falling_new_front: 32
-  falling_new_back: -6
-  falling_new_left: -18
-  falling_new_right: 18
+  falling_thresh_gyro_pitch: 7 # > gyroY
+  falling_thresh_gyro_roll: 7 # > gyroX
+  falling_thresh_orient_roll: 60 # > Point of no return in degrees
+  falling_thresh_orient_pitch: 45 # < Point of no return in degrees
```
[the remark this answers is not in the document]
Comment (SammyRamone) [reply bitbots_hcm/config/hcm_wolfgang.yaml:43]: yes
